Signal/src/util/OWSBackup.m

```diff
@@ -3,964 +3,279 @@
 
-    NSError *error = nil;
-    BOOL success = [[NSFileManager defaultManager] removeItemAtPath:encryptionKeyFilePath error:&error];
-    if (!success || error) {
-        OWSFail(@"%@ could not delete encryption key file: %@", self.logTag, error);
-        return NO;
-    }
-    return YES;
+    [self ensureBackupExportState];
 }
 
-- (BOOL)unzipFilePath
+- (BOOL)shouldHaveBackupExport
 {
-    OWSAssert(self.backupZipPath.length > 0);
-    OWSAssert(self.backupDirPath.length > 0);
-
-    DDLogInfo(@"%@ %s.", self.logTag, __PRETTY_FUNCTION__);
-
-    // Don't use the SSZipArchive convenience methods so that we can add the
-    // encryption key directly as data.
-
-    // TODO: Should we use preserveAttributes?
-    NSError *error = nil;
-    BOOL success = [SSZipArchive unzipFileAtPath:self.backupZipPath
-        toDestination:self.backupDirPath
-        preserveAttributes:YES
-        overwrite:YES
-        nestedZipLevel:0
-        password:self.backupPassword
-        error:&error
-        delegate:self
-        progressHandler:^(NSString *entry, unz_file_info zipInfo, long entryNumber, long total) {
-            DDLogInfo(@"%@ progressHandler: %ld %ld", self.logTag, entryNumber, total);
-
-            CGFloat progress = entryNumber / (CGFloat)total;
-            self.backupProgress = progress;
-        }
-        completionHandler:^(NSString *path, BOOL succeeded, NSError *_Nullable completionError) {
-            DDLogInfo(@"%@ completionHandler: %d %@", self.logTag, succeeded, completionError);
-        }];
-    if (!success || error) {
-        OWSFail(@"%@ failed to unzip file: %@.", self.logTag, error);
+    if (!self.isBackupEnabled) {
         return NO;
     }
-
-    return YES;
-}
-
-- (BOOL)isValidBackup
-{
-    NSString *databasePasswordFilePath =
-        [self.backupDirPath stringByAppendingPathComponent:OWSBackup_DatabasePasswordFilename];
-    if (![[NSFileManager defaultManager] fileExistsAtPath:databasePasswordFilePath]) {
+    if (UIApplication.sharedApplication.applicationState != UIApplicationStateActive) {
+        // Only start backups when app is in the background.
         return NO;
     }
-    NSString *standardUserDefaultsFilePath =
-        [self.backupDirPath stringByAppendingPathComponent:OWSBackup_StandardUserDefaultsFilename];
-    if (![[NSFileManager defaultManager] fileExistsAtPath:standardUserDefaultsFilePath]) {
+    if (![TSAccountManager isRegistered]) {
         return NO;
     }
-    NSString *appUserDefaultsFilePath =
-        [self.backupDirPath stringByAppendingPathComponent:OWSBackup_AppUserDefaultsFilename];
-    if (![[NSFileManager defaultManager] fileExistsAtPath:appUserDefaultsFilePath]) {
+    NSDate *_Nullable lastExportSuccessDate = self.lastExportSuccessDate;
+    NSDate *_Nullable lastExportFailureDate = self.lastExportFailureDate;
+    // Wait N hours before retrying after a success.
+    const NSTimeInterval kRetryAfterSuccess = 24 * kHourInterval;
```

Comment:
There's a lot of constants in this work that are just guesses - they're noted in the code by TODOs or by questions in the design doc.

Reply:
Added to the design doc.